flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSink.java

```diff
@@ -322,7 +323,8 @@ private int deriveSinkParallelism(
                     sinkFunction, env, inputTransform, rowtimeFieldIndex, sinkParallelism);
         } else if (runtimeProvider instanceof SinkProvider) {
             return new SinkTransformation<>(
-                    inputTransform,
+                    createTimestampTransformation(
```

Comment:
I do not think this fully fixes the problem. Currently, users can also use the `DataStreamSinkProvider` in conjunction with the unified sinks i.e. buffered upsert Kafka but in these cases we do not apply the `TimestampOperator`

Reply:
Applied also for `DataStreamSinkProvider`.